src/util/graph.h

```diff
@@ -101,7 +102,19 @@ void intersection(
   }
 }
 
-// a generic graph class with a parametric node type
+/// A generic directed graph with a parametric node type.
+///
+/// The nodes of the graph are stored in the only field of the class, a
+/// std::vector named `nodes`. Nodes are instances of class graph_nodet<E> or a
+/// subclass of it. Graph edges contain a payload object of parametric type E.
+/// By default E is instantiated with an empty class (empty_edget).
+///
+/// Each node is identified by its offset inside the `nodes` vector. The
+/// incomding and outgoing edges of a node are stored as std::maps in the fields
+/// `in` and `out` of the graph_nodet<E>. These maps associate a node identifier
+/// (the offset) to an edge payload (of type E). In fact, observe that two
+/// instances of E are stored per a "mathematical" edge from node A to node B,
```

Comment:
The formal ("mathematical") definition of a directed graph is a pair (V, F) where V is a set of vertices and $F \subseteq V \times V$ are the edges. A "mathematical" edge is an element of F. If you want to label an edge with an object of class `E`, you should associate one such objet to every element of F. However, the maps `in` and `out` in `graph_nodet` associate one object of class `E` to  elements of the set $V \times F$.

As a result, if you have an edge (A,B) in F and you want to label it with an object `e`, the implementation of `graph_nodet` rather stores two copies of `e`, one in `A.out[B]` and another one in `B.in[A]`.

Perhaps I can change "mathematical" to "formal"?